crates/pathfinder/src/p2p_network/sync_handlers.rs

```diff
@@ -21,60 +69,456 @@ pub async fn get_block_headers(
         };
 
-        let transaction_count = tx.transaction_count(header.hash.into())?;
-
-        headers.push(p2p_proto::common::BlockHeader {
-            hash: header.hash.0,
-            parent_hash: header.parent_hash.0,
-            number: header.number.get(),
-            state_commitment: header.state_commitment.0,
-            storage_commitment: header.storage_commitment.0,
-            class_commitment: header.class_commitment.0,
-            sequencer_address: header.sequencer_address.0,
-            timestamp: header.timestamp.get(),
-            gas_price: header.gas_price.0.into(),
-            transaction_count: transaction_count
-                .try_into()
-                .context("Too many transactions")?,
-            transaction_commitment: header.transaction_commitment.0,
-            event_count: 0,
-            event_commitment: header.event_commitment.0,
-            starknet_version: header.starknet_version.take_inner(),
+        let (transactions, receipts) = transactions_and_receipts
+            .into_iter()
+            .map(conv::body::from)
+            .unzip();
+
+        block_bodies.push(p2p_proto::common::BlockBody {
+            transactions,
+            receipts,
         });
 
         count -= 1;
         next_block_number = get_next_block_number(block_number, request.direction);
     }
 
-    Ok(headers)
+    Ok(p2p_proto::sync::BlockBodies { block_bodies })
+}
+
+fn state_diffs(
+    tx: Transaction<'_>,
+    request: p2p_proto::sync::GetStateDiffs,
+) -> anyhow::Result<p2p_proto::sync::StateDiffs> {
+    let mut count = std::cmp::min(request.count, MAX_STATE_UPDATES_COUNT);
+    let mut block_state_updates = Vec::new();
+
+    let mut next_block_number = tx
+        .block_id(BlockHash(request.start_block).into())?
+        .map(|(n, _)| n);
+
+    while let Some(block_number) = next_block_number {
+        if count == 0 {
+            break;
+        }
+
+        let Some(block_hash) = tx.block_id(block_number.into())?.map(|(_, h)| h) else {
+            // No such block
+            break;
+        };
+
+        let Some(state_diff) = tx.state_diff(block_number.into())? else {
+            // No such state update, shouldn't happen with a single source of truth in L2...
+            break;
+        };
+
+        block_state_updates.push(p2p_proto::sync::BlockStateUpdateWithHash {
+            block_hash: block_hash.0,
+            state_update: conv::state_update::from(state_diff),
+        });
+
+        count -= 1;
+        next_block_number = get_next_block_number(block_number, request.direction);
+    }
+
+    Ok(p2p_proto::sync::StateDiffs {
+        block_state_updates,
+    })
+}
+
+fn classes(
+    tx: Transaction<'_>,
+    request: p2p_proto::sync::GetClasses,
+) -> anyhow::Result<p2p_proto::sync::Classes> {
+    let mut classes = Vec::new();
+    for hash in request.class_hashes {
+        let Some(class) = tx.class_definition(ClassHash(hash))? else {
+            break;
```

Comment:
What's the expected behavior for class hashes the node does now know? Currently we're stopping at the first such instance, which might be a bit unoptimal?

Also, class definitions can be large -- what guarantees do we have around max message size?

Reply:
These have to be addressed in the next PR as the current one is already large. I agree bailing out on the first mismatch is suboptimal to say the least. Message size limit is not handled yet.

Reply:
This PR is more like the first stab at the subject on which we could build further.